[FSDP] Include buffers in `ignored_modules`

awgu opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#76784 [FSDP] Include buffers in `ignored_modules`**

This ignores the buffers contained in ignored modules passed into the `ignored_modules` argument for FSDP. For robustness, it also re-generates the parameter and buffer names to ignore on the fly rather than saving them during the constructor.

Differential Revision: [D36120759](https://our.internmc.facebook.com/intern/diff/D36120759)

Commit:
[FSDP] Include buffers in `ignored_modules`

[ghstack-poisoned]

awgu commented:
@awgu has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D36120759).

rohan-varma commented on test/distributed/fsdp/test_fsdp_state_dict.py:
```diff
@@ -440,30 +443,29 @@ def test_wrong_state_dict_config(self):
         # Check that the state dict can be loaded into a non-wrapped version of
         # the model
-        nonwrapped_model = Model(wrap_fsdp=False).cuda()
+        nonwrapped_model = Model(wrap_fsdp=False, register_buffer=True).cuda()
```
Can we also add a check:

Call state_dict() twice, ensure ignored parameters and buffers have the same data_ptr?

rohan-varma commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -768,19 +759,69 @@ def _get_ignored_modules(
+        """
+        ignored_named_tensors = set()
+
+        def f(ignored_named_tensors, module: torch.nn.Module, prefix: str):
```
Can we dedupe this with `def f()` in get_param_to_unflat_param_name? Seems like a fair amount of the code is duplicated but there are some slight differences.

rohan-varma commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -1445,12 +1486,14 @@ def _full_post_state_dict_hook(
-            # sharded
-            clean_param_name = key.replace(FSDP_WRAPPED_MODULE + ".", "").replace(FPW_MODULE + ".", "")
-            if clean_param_name in ignored_param_names:
+            # Do not need to clone ignored parameters and buffers since they
```
Not needed in this PR, but we don't actually even need to clone managed FSDP buffers, because they are not sharded as well

rohan-varma reviewed (approved):
LGTM, just a few minor comments. Thanks for the quick fix!

Also, can we update ignored_modules docstring to mention it includes buffers now if it doesn't already.

Commit:
Update on "[FSDP] Include buffers in `ignored_modules`"


Differential Revision: [D36120759](https://our.internmc.facebook.com/intern/diff/D36120759)

[ghstack-poisoned]

awgu commented:
Failures in `distributed, 1, 2, linux.rocm.gpu` are due to FSDP not supporting Gloo (no `dist._all_gather_base()`), which is unrelated to this PR.

awgu commented:
@pytorchbot merge this please

Commit:
Update on "[FSDP] Include buffers in `ignored_modules`"


Differential Revision: [D36120759](https://our.internmc.facebook.com/intern/diff/D36120759)

[ghstack-poisoned]

awgu commented:
@awgu has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D36120759).

awgu commented:
@pytorchbot merge this please
